### PR TITLE
Push To Master: Parameter Sweep visualization

### PR DESCRIPTION
**Pull request by gwaygenomics**
Three files added in this PR.

1. `scripts/summarize_paramsweep.py` - small python script to aggregate parameter sweep results
2. `results/parameter_sweep_full_results.tsv` - the result of running the above script
3. `scripts/viz/param_sweep_viz.R` - performs all visualizations

### CODE REVIEW

**Comment by gwaygenomics** [comment]
@jaclyn-taroni - here are the scripts to generate the figures given in #51

File: scripts/viz/param_sweep_viz.R [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  subset_param <- param_melt_df %>% dplyr::filter(kappa == k)
+  
+  p <- ggplot(subset_param, aes(x = train_epoch, y = loss)) +
+    geom_line(aes(color = learning_rate, linetype = loss_type), size = 0.5) + 
+    facet_grid(batch_size ~ epochs, scales = "free") + theme_bw() + 
+    ggtitle(paste("kappa =", k))
+  
+  output_fig <- file.path("figures", "param_sweep", "full_param_")
+  output_fig <- paste0(output_fig, k, "_kappa.png")
+  print(p)
```
Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: Do you need `print` here? Would the following work? 

```R
ggsave(output_fig, plot = p, height = 5, width = 6)
```

File: scripts/viz/param_sweep_viz.R [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  subset_param <- param_melt_df %>% dplyr::filter(epochs == e)
+  
+  p <- ggplot(subset_param, aes(x = train_epoch, y = loss)) +
+    geom_line(aes(color = learning_rate, linetype = loss_type), size = 0.5) + 
+    facet_grid(batch_size ~ kappa, scales = "free") + theme_bw() + 
+    ggtitle(paste("epochs =", e))
+  
+  output_fig <- file.path("figures", "param_sweep", "full_param_")
+  output_fig <- paste0(output_fig, e, "_epochs.png")
+  print(p)
```
Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: Similar comment to above

File: scripts/viz/param_sweep_viz.R [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+param_melt_df$batch_size <- factor(param_melt_df$batch_size,
+                                   levels = c(50, 100, 128, 200))
+param_melt_df$epochs <- factor(param_melt_df$epochs,
+                               levels = c(10, 25, 50, 100))
+
+# Sweep over measurements of kappa with batch_size and epochs as facets
+for (k in unique(param_melt_df$kappa)) {
+  subset_param <- param_melt_df %>% dplyr::filter(kappa == k)
+  
+  p <- ggplot(subset_param, aes(x = train_epoch, y = loss)) +
```
Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: So I think these are the plots included in [`parameter_sweep.md`](https://github.com/greenelab/tybalt/blob/master/parameter_sweep.md) - is that correct? The fact that the "row" panels are batch sizes is not labeled in those plots and is not immediately obvious in the absence of a figure legend. I might recode these so it's clear what these facets are showing. For instance:

```R
param_melt_df$batch_size <- 
  factor(param_melt_df$batch_size, levels = c(50, 100, 128, 200))
param_melt_df$batch_size <- 
  dplyr::recode_factor(param_melt_df$batch_size, 
                       `50` = "batch: 50", 
                       `100` = "batch: 100",
                       `128` = "batch: 128",
                       `200` = "batch: 200") 
```

Might even do something similar for the epochs, but at least the x-axis label `train_epoch` is a pointer for what's going on.

File: scripts/viz/param_sweep_viz.R [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# Pan-Cancer Variational Autoencoder
+# Gregory Way 2017
+#
+# Visualizing results of parameter sweep
+
+library(readr)
+library(ggplot2)
```
Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: Is there a reason you're opting to load these packages here rather than going with the double colon nomenclature? Can be a little clunky for plotting certainly.

File: scripts/viz/param_sweep_viz.R [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  output_fig <- paste0(output_fig, k, "_kappa.png")
+  print(p)
+  ggsave(output_fig, height = 5, width = 6)
+}
+
+# Sweep over epochs with batch_size and kappa as facets
+for (e in unique(param_melt_df$epochs)) {
+  subset_param <- param_melt_df %>% dplyr::filter(epochs == e)
+  
+  p <- ggplot(subset_param, aes(x = train_epoch, y = loss)) +
```
Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: I have the same comment about the lack of labeling for the facets. I am a little less concerned here because they are not included in markdown documentation w/o a figure legend.

File: scripts/viz/param_sweep_viz.R [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  output_fig <- file.path("figures", "param_sweep", "full_param_")
+  output_fig <- paste0(output_fig, e, "_epochs.png")
+  print(p)
+  ggsave(output_fig, height = 5, width = 6)
+}
+
+final_select_df <- param_melt_df %>% filter(loss_type == "val_loss")
+final_select_df <- final_select_df %>%
+  group_by(learning_rate, batch_size, epochs) %>%
+  summarize(min_loss = min(loss))
```
Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: This `min_loss` should be the value for the final epoch but not necessarily, correct? I think there could be a little more information around this figure in [`parameter_sweep.md`](https://github.com/greenelab/tybalt/blob/master/parameter_sweep.md).

File: scripts/viz/param_sweep_viz.R (hunk 4 above)
Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: yeah, must be left over from prototyping. I still need `dplyr` loaded for piping though

Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: Can probably get away with
```R
`%>%` <- dplyr::`%>%`
```

Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: yeah, looks like I can - although I haven't seen that before.

Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: also, unless you've come across sound reasons to use `ggplot2` namespace calls, I am going to leave loading in the library. 

I think that

```R
ggplot(subset_param, aes(x = train_epoch, y = loss)) +
    geom_line(aes(color = learning_rate, linetype = loss_type), size = 0.5) + 
    facet_grid(batch_size ~ epochs, scales = "free") + theme_bw() + 
    ggtitle(paste("kappa =", k))
```

is easier to digest than

```R
ggplot2::ggplot(subset_param, ggplot2::aes(x = train_epoch, y = loss)) +
    ggplot2::geom_line(ggplot2::aes(color = learning_rate, linetype = loss_type), size = 0.5) + 
    ggplot2::facet_grid(batch_size ~ epochs, scales = "free") + ggplot2::theme_bw() + 
    ggplot2::ggtitle(paste("kappa =", k))
```

Comment (jaclyn-taroni) [inline scripts/viz/param_sweep_viz.R]: I believe I learned the pipe trick from our own esteemed @dhimmel

Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: me too! There's nothing in the style guide about it

It definitely makes sense to do if there are functions of the same name in different packages, and also if you only need single functions in a large library. But, given the simplicity of this script, I think I will stay with the call to `library(ggplot2)`. I will however, remove the calls to `library(dplyr)` and `library(readr)`. Change in next commit!

File: scripts/viz/param_sweep_viz.R (hunk 3 above)
Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: awesome! Yes, I will do this

File: scripts/viz/param_sweep_viz.R (hunk 1 above)
Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: nope, again, not removed after prototyping. change coming!

File: scripts/viz/param_sweep_viz.R (hunk 5 above)
Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: For consistency, I will include the change

File: scripts/viz/param_sweep_viz.R (hunk 6 above)
Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: great catch! I will update this to take the last value

Comment (gwaygenomics) [inline scripts/viz/param_sweep_viz.R]: Rather than update `parameter_sweep.md` - I updated the axis label. Indeed, the bug fix does alter the figure a bit

**Comment by gwaygenomics** [comment]
Thanks for the thorough review @jaclyn-taroni - I believe I addressed all of the comments. I also included updated plots in this PR
